### PR TITLE
Abstracting out the liquidator logic

### PR DESCRIPTION
**Pull request by IanWoodard**
Abstracting out the liquidator logic to enable the deployment of bots on multiple chains at once. While abstracting out the logic, I also cleaned up the code to improve error handling and add more details to logs. Additionally, I added a library to spread out the calls to avoid being rate-limited. Added logic in the liquidator to check that it hasn't errored too many times, which is incorporated into the `isHealthy` check. Additionally, I added more explicit messages to the health check to make it easier to debug.

### CODE REVIEW

File: app/src/Liquidator.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+import Bottleneck from "bottleneck";
+
+const FACTORY_ADDRESS: string = process.env.FACTORY_ADDRESS!;
+const CREATE_ACCOUNT_TOPIC_ID: string = process.env.CREATE_ACCOUNT_TOPIC_ID!;
+const WALLET_ADDRESS = process.env.WALLET_ADDRESS!;
+const ALOE_INITIAL_DEPLOY = 0;
+const POLLING_INTERVAL = 60_000;
+const STATUS_HEALTHY = 200;
+const STATUS_NOT_HEALTHY = 503;
+const REQUEST_RATE_LIMIT = 250;
```
Comment (aneesh1) [inline app/src/Liquidator.ts:17]: Units? Based on the [docs](https://www.npmjs.com/package/bottleneck#schedule), perhaps `REQUEST_RATE_LIMIT_IN_MS`?

File: app/src/Liquidator.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+import LiquidatorABIJson from "./abis/Liquidator.json";
+import TxManager from "./TxManager";
+import winston from "winston";
+import Bottleneck from "bottleneck";
+
+const FACTORY_ADDRESS: string = process.env.FACTORY_ADDRESS!;
+const CREATE_ACCOUNT_TOPIC_ID: string = process.env.CREATE_ACCOUNT_TOPIC_ID!;
+const WALLET_ADDRESS = process.env.WALLET_ADDRESS!;
+const ALOE_INITIAL_DEPLOY = 0;
+const POLLING_INTERVAL = 60_000;
```
Comment (aneesh1) [inline app/src/Liquidator.ts:14]: Units?

**Review by aneesh1 — APPROVED** [review]
A couple minor comments - looks good!

File: app/src/Liquidator.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    );
+  }
+
+  private scan(borrowers: string[]): void {
+    borrowers.forEach((borrower) => {
+      this.limiter.schedule(async () => {
+        const solvent: boolean = await this.isSolvent(borrower);
+        console.log("Is solvent?", solvent, borrower);
+        if (!solvent) {
+          // TODO: We probably don't actually want to log this here, at least not at "info" level (since that'll send it to Slack every time).
```
[the remark this answers is not in the document]
Comment (IanWoodard) [reply app/src/Liquidator.ts:167]: That was already there

File: app/src/Liquidator.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+      },
+      reconnect: {
+        auto: true,
+        delay: 5000,
+        maxAttempts: 5,
+        onTimeout: false,
+      },
+    });
+    this.web3 = new Web3(provider);
+    this.web3.eth.handleRevert = true;
```
[the remark this answers is not in the document]
Comment (IanWoodard) [reply app/src/Liquidator.ts:51]: Previously we could not create separate instances of liquidators. Now that we are adding that capability, we must create separate providers for each instance, as instances are used for different chains requiring different `jsonRpcURL`s. There may be some confusion stemming from us only having one liquidator currently, and thus only on `jsonRPCURL`, but I assure you we will have more in the future.

[the remark this answers is not in the document]
Comment (aneesh1) [reply app/src/Liquidator.ts:51]: Ah - my mistake!

File: app/src/Liquidator.ts [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+  private collectBorrowers(block: number) {
+    this.web3.eth.subscribe(
+      "logs",
+      {
+        address: FACTORY_ADDRESS,
+        topics: [CREATE_ACCOUNT_TOPIC_ID],
+        fromBlock: block,
+      },
+      this.collectBorrowersCallback.bind(this)
```
Comment (haydenshively) [inline app/src/Liquidator.ts:160]: Good call on the `.bind(this)` -- I had a lot of trouble with forgetting that in my old bots.

**Review by haydenshively — APPROVED** [review]
This is a pretty substantial architectural change. Were you able to test this locally? If not, I think we should run it in parallel with the old one for Optimism, and test this one out on Arbitrum.

File: app/src/index.ts [hunk 6]
```diff
@@ -1,219 +1,106 @@
...
+    liquidators.map(async (liquidator) => {
+      const healthCheckResponse: HealthCheckResponse =
+        await liquidator.isHealthy();
+      if (healthCheckResponse.code !== STATUS_OK) {
+        return res
+          .status(healthCheckResponse.code)
+          .send({ error: healthCheckResponse.message });
+      }
+    })
+  );
```
[the remark this answers is not in the document]
Comment (IanWoodard) [reply app/src/index.ts:44]: It is mostly the same, just called on each instance of the liquidator. I also added another check that ensures we haven't errored too many times in the liquidator. Additionally, I added more explicit messages to help debug down the line if anything goes wrong.

**Comment by IanWoodard** [comment]
> This is a pretty substantial architectural change. Were you able to test this locally? If not, I think we should run it in parallel with the old one for Optimism, and test this one out on Arbitrum.

I have tested it a fair bit locally. I am fairly confident that it will work seamlessly, but if anything does go wrong, we will know pretty easily.

File: app/src/TxManager.ts [hunk 7]
```diff
@@ -24,34 +21,41 @@ type LiquidationTxInfo = {
     retries: number;
 }
 
+const MAX_GAS_PRICE_FOR_CHAIN: Map<number, string> = new Map<number, string>([
+    [1, "100000000000"],
+    [5, "10000000000"],
+    [10, "10000000"],
+    [42161, "1000000000"],
```
Comment (haydenshively) [inline app/src/TxManager.ts:28]: These would be more readable if you used web3.js conversion tools (from gwei --> wei)

**Review by haydenshively — APPROVED** [review]

**Pull request merged**